accepted/2021/default_string_globalization.md

```diff
@@ -0,0 +1,390 @@
+
+[Java 16's `String` class](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/String.html) (which is the closest to .NET's `string` class) is a good example of this principle. Per their own documentation:
+
+> Unless otherwise noted, methods for comparing Strings do not take locale into account.
+
+And in fact, the _only_ two Java `String` APIs that are culture-aware are case-conversion routines.
+
+```java
+String s = getString();
+
+// parameterless overloads use current locale by default
+String s1 = s.toUpperCase();
+String s2 = s.toLowerCase();
+
+// or you can pass a Locale instance if you don't want to use the current locale
+// Java's Locale.ROOT is roughly equivalent to .NET's CultureInfo.InvariantCulture
+String s3 = s.toUpperCase(Locale.ROOT);
+String s4 = s.toLowerCase(Locale.ROOT);
+```
+
+And even though these are the only two linguistic-by-default APIs on Java's `String` type, problematic usage of them is still so widespread that CERT's coding rules for Java [forbid calling the parameterless overloads](https://wiki.sei.cmu.edu/confluence/display/java/STR02-J.+Specify+an+appropriate+locale+when+comparing+locale-dependent+data). This CERT rule STR02-J is roughly equivalent to .NET's [CA1310 rule](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1310), which flags certain overloads of `string` APIs as problematic.
+
+Put another way, the second characteristic mentioned above might be paraphrased as "developers shouldn't be exposed to globalization-related concerns unless they have imported the _System.Globalization_ namespace into their file."
+
+## Proposed solution
+
+To solve this problem for the .NET ecosystem at large, we propose to take the following actions on `string`.
+
+1. Obsolete (as warning) all linguistic-by-default methods on `string` with a new __SYSLIBxxxx__ identifier.
+2. Mark all newly-obsoleted methods as `[EditorBrowsable(EditorBrowsableState.Never)]`.
+3. Mark all methods which take a `CultureInfo` parameter as `[EditorBrowsable(EditorBrowsableState.Never)]`, but do not obsolete them.
+4. Add simple ordinal overloads for all obsoleted methods. (These overloads _should not_ take `StringComparison` as a parameter. More on this later.)
+5. Add simple "ignore case" ordinal overloads for all methods which currently don't have them.
+6. Create fixers to migrate call sites from the now-obsoleted methods onto newer methods.
+
+And we propose to take the following actions on `string`-like APIs.
+
+1. Mark all linguistic-related APIs as `[EditorBrowsable(EditorBrowsableState.Never)]`, but do not obsolete them. (Exception: do not mark APIs if they're in the _System.Globalization_ namespace.)
+2. Create analyzers to detect `Comparer<string>.Default`, `new SortedSet<string>()`, and similar code patterns that implicitly rely on the current culture.
+
+### List of API obsoletions and additions
```

Comment:
Are we going to release an updated version of netstandard or are the developers targetting multiple versions going to be faced with warnings everywhere?

Reply:
I think the challenge here is we are going to have a new APIs which will be in .NET 6.0 and up. I don't think we can add these to the netstandard. 

CC @terrajobst

Reply:
These wouldn't be in standard. If you're multi-targeting, I believe per the way this proposal is written, the only methods that would be marked `[Obsolete]` are methods that the code analyzers CA1307 and CA1310 would have flagged anyway as likely incorrect usage. A significant number of methods would be marked EB-never but wouldn't produce warnings.